extensions/ql-vscode/src/local-queries.ts

```diff
@@ -7,384 +7,575 @@ import {
+        // exception, and by returning a result with a failure code. This is how the code worked
+        // before the refactoring, so it's been preserved, but we should probably figure out how
+        // to unify both error handling paths.
+        const err = asError(e);
+        await localQueryRun.fail(err);
+        throw e;
+      }
     } finally {
-      await qhm.refreshTreeView();
       source.dispose();
     }
   }
-}
 
-interface DatabaseQuickPickItem extends QuickPickItem {
-  databaseItem: DatabaseItem;
-}
-
-async function compileAndRunQueryOnMultipleDatabases(
-  cliServer: CodeQLCliServer,
-  qs: QueryRunner,
-  qhm: QueryHistoryManager,
-  dbm: DatabaseManager,
-  databaseUI: DatabaseUI,
-  localQueryResultsView: ResultsView,
-  queryStorageDir: string,
-  progress: ProgressCallback,
-  token: CancellationToken,
-  uri: Uri | undefined,
-): Promise<void> {
-  let filteredDBs = dbm.databaseItems;
-  if (filteredDBs.length === 0) {
-    void showAndLogErrorMessage(
-      "No databases found. Please add a suitable database to your workspace.",
-    );
-    return;
-  }
-  // If possible, only show databases with the right language (otherwise show all databases).
-  const queryLanguage = await findLanguage(cliServer, uri);
-  if (queryLanguage) {
-    filteredDBs = dbm.databaseItems.filter(
-      (db) => db.language === queryLanguage,
-    );
+  private async compileAndRunQueryOnMultipleDatabases(
+    progress: ProgressCallback,
+    token: CancellationToken,
+    uri: Uri | undefined,
+  ): Promise<void> {
+    let filteredDBs = this.databaseManager.databaseItems;
     if (filteredDBs.length === 0) {
       void showAndLogErrorMessage(
-        `No databases found for language ${queryLanguage}. Please add a suitable database to your workspace.`,
+        "No databases found. Please add a suitable database to your workspace.",
       );
       return;
     }
-  }
-  const quickPickItems = filteredDBs.map<DatabaseQuickPickItem>((dbItem) => ({
-    databaseItem: dbItem,
-    label: dbItem.name,
-    description: dbItem.language,
-  }));
-  /**
-   * Databases that were selected in the quick pick menu.
-   */
-  const quickpick = await window.showQuickPick<DatabaseQuickPickItem>(
-    quickPickItems,
-    { canPickMany: true, ignoreFocusOut: true },
-  );
-  if (quickpick !== undefined) {
-    // Collect all skipped databases and display them at the end (instead of popping up individual errors)
-    const skippedDatabases = [];
-    const errors = [];
-    for (const item of quickpick) {
-      try {
-        await compileAndRunQuery(
-          qs,
-          qhm,
-          databaseUI,
-          localQueryResultsView,
-          queryStorageDir,
-          false,
-          uri,
-          progress,
-          token,
-          item.databaseItem,
+    // If possible, only show databases with the right language (otherwise show all databases).
```

Comment:
Is this comment incorrect? It looks like we show an error if there are no suitable databases found. I'm not sure we want to show all databases.

Reply:
I think the `findLanguage()` call below can return `undefined` if the user cancels the prompt to select a language, and in that case the filtering by language (and the error message) won't happen.